### PR TITLE
K8SPS-20: Rework deployment

### PR DESCRIPTION
**Pull request by egegunes**
New versions of Operator SDK is more aligned with kubebuilder.
Kubebuilder creates YAML config for installing CRDs, RBAC and operator
deployment under `config/` directory. It uses kustomize and with the
help of markes, it automatically updates CRD, RBAC etc. Rather than
breaking this setup, we integrated it with our approach in our other
operators. We have a `deploy/` directory that contains `cr.yaml`,
`crd.yaml`, `rbac.yaml`, `operator.yaml` and these files are generated
by kustomize from `config/`. **If you want to change something in the
manifest, you need to change files in `config/`. Files in `deploy/` will
be overwritten by make targets.**

We have a `Makefile` that helps maintaining manifests and install
CRDs/operator. Basic commands are the following:

1. `make manifests`: Create `crd.yaml`, `rbac.yaml` and `operator.yaml` and
put them under `deploy/`.
2. `make install`: Apply `crd.yaml` and `rbac.yaml` to Kubernetes.
4. `make build`: Build the docker image for operator.
3. `make deploy`: Apply `operator.yaml` to Kubernetes.

We're using `perconalab/percona-server-mysql-operator:main` as our
default image. If you want to override the tag for development:

```
$ make deploy VERSION=k8sps-20
```

If you want to override the image:

```
$ make manifest IMG=percona/percona-server-mysql-operator:0.1.0
```

### CODE REVIEW

File: deploy/operator.yaml [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+      labels:
+        control-plane: percona-server-for-mysql-operator
+    spec:
+      containers:
+      - args:
+        - --leader-elect
+        command:
+        - /usr/local/bin/percona-server-mysql-operator
+        env:
+        - name: WATCH_NAMESPACE
```
Comment (hors) [inline deploy/operator.yaml:39]: Should we support cluster wide from very beginning or not? Also please create the bundle.yaml file

**Review by hors — CHANGES_REQUESTED** [review]

File: deploy/operator.yaml [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+  template:
+    metadata:
+      labels:
+        control-plane: percona-server-for-mysql-operator
+    spec:
+      containers:
+      - args:
+        - --leader-elect
+        command:
+        - /usr/local/bin/percona-server-mysql-operator
```
Comment (hors) [inline deploy/operator.yaml:37]:
```suggestion
        - /usr/local/bin/percona-server-for-mysql-operator
```

**Review by hors — CHANGES_REQUESTED** [review]

File: deploy/operator.yaml (hunk 1 above)
Comment (egegunes) [reply deploy/operator.yaml:39]: https://github.com/percona/percona-server-mysql-operator/pull/12/commits/274cac3208f0d261d435eca72f7d7058f98fbf4d

File: deploy/operator.yaml (hunk 2 above)
Comment (egegunes) [reply deploy/operator.yaml]: https://github.com/percona/percona-server-mysql-operator/pull/12/commits/274cac3208f0d261d435eca72f7d7058f98fbf4d

File: config/rbac/role_binding.yaml [hunk 3]
```diff
@@ -1,12 +1,11 @@
 apiVersion: rbac.authorization.k8s.io/v1
-kind: ClusterRoleBinding
+kind: RoleBinding
```
Comment (delgod) [inline config/rbac/role_binding.yaml:2]: ClusterRoleBinding used with ClusterRole
RoleBinding used with Role
please check `config/rbac/role.yaml`

File: deploy/bundle.yaml [hunk 4]
```diff
@@ -0,0 +1,4271 @@
...
+                                      type: string
+                                    httpHeaders:
+                                      description: Custom headers to set in the request. HTTP allows repeated headers.
+                                      items:
+                                        description: HTTPHeader describes a custom header to be used in HTTP probes
+                                        properties:
+                                          name:
+                                            description: The header field name
+                                            type: string
+
```
Comment (delgod) [inline deploy/bundle.yaml:4196]: ClusterRoleBinding can bind ClusterRole

File: Makefile [hunk 5]
```diff
@@ -93,34 +92,33 @@ vet: ## Run go vet against code.
...
-deploy: manifests kustomize ## Deploy controller to the K8s cluster specified in ~/.kube/config.
-	cd config/manager && $(KUSTOMIZE) edit set image controller=${IMG}
-	$(KUSTOMIZE) build config/default | kubectl apply -f -
+deploy: manifests ## Deploy operator
+	kubectl apply -f $(DEPLOYDIR)/operator.yaml
 
-undeploy: ## Undeploy controller from the K8s cluster specified in ~/.kube/config.
-	$(KUSTOMIZE) build config/default | kubectl delete -f -
+undeploy: ## Undeploy operator
+	kubectl delete -f $(DEPLOYDIR)/operator.yaml
```
Comment (delgod) [inline Makefile:122]: how DEPLOYDIR content is generated? can we add a section here for it?

File: deploy/operator.yaml [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+    kind: ControllerManagerConfig
+    health:
+      healthProbeBindAddress: :8081
+    metrics:
+      bindAddress: 127.0.0.1:8080
+    webhook:
+      port: 9443
+    leaderElection:
+      leaderElect: true
+      resourceName: 08db0feb.percona.com
```
Comment (delgod) [inline deploy/operator.yaml:14]: how resourceName is selected?

**Review by hors — CHANGES_REQUESTED** [review]

File: Makefile [hunk 7]
```diff
@@ -3,7 +3,8 @@
 # To re-generate a bundle for another specific version without changing the standard setup, you can:
 # - use the VERSION as arg of the bundle target (e.g make bundle VERSION=0.0.2)
 # - use environment variables to overwrite this value (e.g export VERSION=0.0.2)
-VERSION ?= 0.0.1
+NAME ?= percona-server-for-mysql-operator
+VERSION ?= main
```
[the remark this answers is not in the document]
Comment (egegunes) [reply Makefile]: https://github.com/percona/percona-server-mysql-operator/pull/12/commits/d58f477d1e45be115621da452a9b482e4aeb4cd6

File: README.md [hunk 8]
```diff
@@ -1,12 +1,18 @@
-# Percona Server Operator
+# Percona Server for MySQL Operator
```
[the remark this answers is not in the document]
Comment (egegunes) [reply README.md]: https://github.com/percona/percona-server-mysql-operator/pull/12/commits/d58f477d1e45be115621da452a9b482e4aeb4cd6

File: config/rbac/role_binding.yaml (hunk 3 above)
Comment (egegunes) [reply config/rbac/role_binding.yaml]: https://github.com/percona/percona-server-mysql-operator/pull/12/commits/b52deb68ebc8a0fd724890bc0600115eae1aa652

File: deploy/bundle.yaml (hunk 4 above)
Comment (egegunes) [reply deploy/bundle.yaml]: https://github.com/percona/percona-server-mysql-operator/pull/12/commits/b52deb68ebc8a0fd724890bc0600115eae1aa652

File: Makefile (hunk 5 above)
Comment (egegunes) [reply Makefile:123]: https://github.com/percona/percona-server-mysql-operator/pull/12/commits/d58f477d1e45be115621da452a9b482e4aeb4cd6

File: deploy/bundle.yaml [hunk 9]
```diff
@@ -0,0 +1,4271 @@
+apiVersion: apiextensions.k8s.io/v1
```
[the remark this answers is not in the document]
Comment (egegunes) [reply deploy/bundle.yaml:1]: https://github.com/percona/percona-server-mysql-operator/pull/12/commits/d58f477d1e45be115621da452a9b482e4aeb4cd6

File: deploy/operator.yaml (hunk 6 above)
Comment (egegunes) [reply deploy/operator.yaml:14]: I believe it's generated automatically

File: config/rbac/leader_election_role_binding.yaml [hunk 10]
```diff
@@ -1,12 +1,11 @@
 apiVersion: rbac.authorization.k8s.io/v1
-kind: RoleBinding
+kind: ClusterRoleBinding
```
Comment (delgod) [inline config/rbac/leader_election_role_binding.yaml:2]: why we need to use ClusterRole and ClusterRoleBinding for leader-election? it is namespaced operation

File: deploy/operator.yaml (hunk 6 above)
Comment (delgod) [reply deploy/operator.yaml:14]: up to @hors

File: deploy/rbac.yaml [hunk 11]
```diff
@@ -0,0 +1,198 @@
...
+  - get
+  - list
+  - patch
+  - update
+  - watch
+- apiGroups:
+  - apps
+  resources:
+  - deployments
+  - replicasets
```
Comment (delgod) [inline deploy/rbac.yaml:83]: do we work with replicasets?

File: deploy/rbac.yaml [hunk 12]
```diff
@@ -0,0 +1,198 @@
...
+  - delete
+  - get
+  - list
+  - patch
+  - update
+  - watch
+- apiGroups:
+  - batch
+  resources:
+  - cronjobs
```
Comment (delgod) [inline deploy/rbac.yaml:96]: do we work with cronjobs?

File: config/rbac/leader_election_role_binding.yaml (hunk 10 above)
Comment (egegunes) [reply config/rbac/leader_election_role_binding.yaml:2]: kubebuilder generates this automatically. please see the https://github.com/percona/percona-server-mysql-operator/blob/2c86dbff36e90d558e8280b5fded892aff21449a/deploy/rbac.yaml for the end result

File: deploy/rbac.yaml [hunk 13]
```diff
@@ -0,0 +1,198 @@
...
+  - list
+  - patch
+  - update
+  - watch
+- apiGroups:
+  - ps.percona.com
+  resources:
+  - perconaserverformysqlbackups
+  - perconaserverformysqlbackups/finalizers
+  - perconaserverformysqlbackups/status
```
Comment (delgod) [inline deploy/rbac.yaml:137]: why we need separate sections for backups and restores and ps itself? is it automatically generated?

**Review by delgod — CHANGES_REQUESTED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |

commit: https://github.com/percona/percona-server-mysql-operator/pull/12/commits/4e6fd8367be4da6cee9bc0adfa5f4de4e01ec1d5
image: `perconalab/percona-server-mysql-operator:PR-12-4e6fd83`

File: deploy/rbac.yaml (hunk 13 above)
Comment (egegunes) [reply deploy/rbac.yaml]: all these files are automatically generated

File: deploy/rbac.yaml (hunk 12 above)
Comment (egegunes) [reply deploy/rbac.yaml]: https://github.com/percona/percona-server-mysql-operator/pull/12/commits/def7c33fb4066ac75217d361b878be9bf170fcca

File: deploy/rbac.yaml (hunk 11 above)
Comment (egegunes) [reply deploy/rbac.yaml]: https://github.com/percona/percona-server-mysql-operator/pull/12/commits/def7c33fb4066ac75217d361b878be9bf170fcca

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |

commit: https://github.com/percona/percona-server-mysql-operator/pull/12/commits/def7c33fb4066ac75217d361b878be9bf170fcca
image: `perconalab/percona-server-mysql-operator:PR-12-def7c33`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |

commit: https://github.com/percona/percona-server-mysql-operator/pull/12/commits/4f42ba6720d549a775760fa3b0c56b56793cd7f3
image: `perconalab/percona-server-mysql-operator:PR-12-4f42ba6`

File: build/ps-entrypoint.sh [hunk 14]
```diff
@@ -126,7 +126,7 @@ _get_tmpdir() {
 
 	tmpdir_path=$(_get_cnf_config mysqld tmpdir "")
 	if [[ -z ${tmpdir_path} ]]; then
-		tmpdir_path=$(_get_cnf_config xtrabackup tmpdir "")
+		tmpdir_path=$(_get_cnf_config mysqld tmpdir "")
```
Comment (delgod) [inline build/ps-entrypoint.sh:129]: why this change needed?

File: cmd/bootstrap/main.go [hunk 15]
```diff
@@ -216,14 +225,31 @@ func getTopology(peers sets.String) (string, []string, error) {
...
 
+		replicaHost, err := db.ReportHost()
+		if err != nil {
+			return "", nil, errors.Wrap(err, "get report_host")
+		}
+		if replicaHost == "" {
+			continue
+		}
+
+		if peers.Len() > 1 {
```
Comment (delgod) [inline cmd/bootstrap/main.go:236]: why do we need to check the number of alive peers?

File: pkg/database/mysql/env.go [hunk 16]
```diff
@@ -10,6 +10,10 @@ func (m *MySQL) env() []corev1.EnvVar {
 			Name:  "SERVICE_NAME",
 			Value: m.ServiceName(),
 		},
+		{
+			Name:  "SERVICE_NAME_UNREADY",
```
Comment (delgod) [inline pkg/database/mysql/env.go:14]: where do we use this service name?

**Review by delgod — CHANGES_REQUESTED** [review]

File: pkg/database/mysql/env.go (hunk 16 above)
Comment (egegunes) [reply pkg/database/mysql/env.go:14]: https://github.com/percona/percona-server-mysql-operator/pull/12/files#diff-76e49ba3fd4f13d43ef2fc0f5461e24b60fccddbc9116e758bd3960dce6de0e6R47

File: cmd/bootstrap/main.go (hunk 15 above)
Comment (egegunes) [reply cmd/bootstrap/main.go:236]: @hors ^^

File: build/ps-entrypoint.sh (hunk 14 above)
Comment (egegunes) [reply build/ps-entrypoint.sh:129]: @hors ^^

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |

commit: https://github.com/percona/percona-server-mysql-operator/pull/12/commits/38d6e66842d97b3e5f4e9039385f502163d0a195
image: `perconalab/percona-server-mysql-operator:PR-12-38d6e66`

File: build/ps-entrypoint.sh (hunk 14 above)
Comment (hors) [reply build/ps-entrypoint.sh]: it was misprint, I have fixed it 
@delgod  thank's

File: cmd/bootstrap/main.go (hunk 15 above)
Comment (delgod) [reply cmd/bootstrap/main.go:236]: @hors don't have the explanation why this line (`if peers.Len() >1 {`) is needed

**Review by delgod — CHANGES_REQUESTED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |

commit: https://github.com/percona/percona-server-mysql-operator/pull/12/commits/bf98e315fa14d9f1d74d9e8d3e55f7338add2d01
image: `perconalab/percona-server-mysql-operator:PR-12-bf98e31`

**Review by delgod — APPROVED** [review]

**Review by hors — APPROVED** [review]
